llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,302 @@
+  }
+
+  /// Iterator on all the elements of the vector
+  /// which have actually being constructed.
+  class MaterialisedIterator {
+    const PagedVector *PV;
+    size_t ElementIdx;
+
+  public:
+    using iterator_category = std::forward_iterator_tag;
+    using value_type = T;
+    using difference_type = std::ptrdiff_t;
+    using pointer = T *;
+    using reference = T &;
+
+    MaterialisedIterator(PagedVector const *PV, size_t ElementIdx)
+        : PV(PV), ElementIdx(ElementIdx) {}
+
+    /// Pre-increment operator.
+    ///
+    /// When incrementing the iterator, we skip the elements which have not
+    /// been materialised yet.
+    MaterialisedIterator &operator++() {
+      ++ElementIdx;
+      if (ElementIdx % PageSize == 0) {
+        while (ElementIdx < PV->Size &&
+               PV->PageToDataPtrs[ElementIdx / PageSize] == InvalidPage)
+          ElementIdx += PageSize;
+        if (ElementIdx > PV->Size)
+          ElementIdx = PV->Size;
+      }
+
+      return *this;
+    }
+
+    MaterialisedIterator operator++(int) {
+      MaterialisedIterator Copy = *this;
+      ++*this;
+      return Copy;
+    }
+
+    /// Dereference operator.
+    ///
+    /// Notice this can materialise elements if needed so there might be
+    /// a page allocation and additional construction of the elements of
+    /// such page.
+    T const &operator*() const {
+      assert(ElementIdx < PV->Size);
+      assert(PV->PageToDataPtrs[ElementIdx / PageSize] != InvalidPage);
+      T *PagePtr = PV->PageToDataPtrs[ElementIdx / PageSize];
+      return PagePtr[ElementIdx % PageSize];
+    }
+
+    friend bool operator==(MaterialisedIterator const &LHS,
+                           MaterialisedIterator const &RHS);
+    friend bool operator!=(MaterialisedIterator const &LHS,
+                           MaterialisedIterator const &RHS);
+
+    [[nodiscard]] size_t getIndex() const { return ElementIdx; }
+  };
+
+  /// Equality operator.
+  friend bool operator==(MaterialisedIterator const &LHS,
+                         MaterialisedIterator const &RHS) {
+    assert(LHS.PV == RHS.PV);
+    // Make sure we are comparing either end iterators or iterators pointing
+    // to materialised elements.
+    // It should not be possible to build two iterators pointing to non
+    // materialised elements.
+    assert(LHS.ElementIdx >= LHS.PV->Size ||
+           (LHS.ElementIdx / PageSize < LHS.PV->PageToDataPtrs.size() &&
+            LHS.PV->PageToDataPtrs[LHS.ElementIdx / PageSize] != InvalidPage));
+    assert(RHS.ElementIdx >= RHS.PV->Size ||
+           (RHS.ElementIdx / PageSize < RHS.PV->PageToDataPtrs.size() &&
+            RHS.PV->PageToDataPtrs[RHS.ElementIdx / PageSize] != InvalidPage));
+    return LHS.ElementIdx == RHS.ElementIdx;
+  }
+
+  friend bool operator!=(MaterialisedIterator const &LHS,
+                         MaterialisedIterator const &RHS) {
+    return !(LHS == RHS);
+  }
+
+  friend class MaterialisedIterator;
```

Comment:
It's generally not necessary to befriend a member class; members (including member classes) have access to the private members of the enclosing class.
```suggestion
```